File: packages/provide-uterm/src/provide/uterm/control/plane/memory/engine.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from provide.uterm.control.plane.memory.approval_store import MemoryApprovalStore
from provide.uterm.control.plane.memory.graphical_target_store import MemoryGraphicalTargetStore
from provide.uterm.control.plane.memory.lease_store import MemoryLeaseStore
from provide.uterm.control.plane.memory.session_store import MemorySessionStore
from provide.uterm.control.plane.memory.token_store import MemoryTokenStore
from provide.uterm.control.plane.memory.transaction import MemoryState, MemoryTransaction

if TYPE_CHECKING:
    from provide.uterm.control.plane.capability import EngineCapabilities
    from provide.uterm.control.plane.types import ControlPlaneConfig


@dataclass(slots=True)
class MemoryControlPlane:
    """In-memory control-plane backend with shared mutable state."""

    config: ControlPlaneConfig
    capabilities: EngineCapabilities = field(init=False)
    _state: MemoryState = field(init=False, repr=False)
    _lock: asyncio.Lock = field(init=False, repr=False)

    def __post_init__(self) -> None:
        self.capabilities = self.config.capabilities
        self._state = MemoryState()
        self._lock = asyncio.Lock()

    async def open(self) -> None:
        return None

    async def close(self) -> None:
        return None

    async def migrate(self) -> None:
        return None

    async def begin(self) -> MemoryTransaction:
        return MemoryTransaction(self._state, self._lock)

    async def reap(self, *, now: float, retention_s: int) -> int:
        """Physically drop control-plane rows whose soft-delete/expiry timestamp
        is older than ``now - retention_s``.  Mirrors the SQLite reap predicate
        semantics (strict ``<``, IS-NOT-NULL guards) so both backends prune the
        same rows.  Returns the total number of records removed."""
        cutoff = now - retention_s
        async with self._lock:
            s = self._state
            before = len(s.resume_tokens) + len(s.session_tokens) + len(s.sessions) + len(s.leases) + len(s.approvals)
            s.resume_tokens = {
                k: r
                for k, r in s.resume_tokens.items()
                if not ((r.revoked_at is not None and r.revoked_at < cutoff) or r.expires_at < cutoff)
            }
            s.session_tokens = {
                k: r
                for k, r in s.session_tokens.items()
                if not (
                    (r.revoked_at is not None and r.revoked_at < cutoff)
                    or (r.expires_at is not None and r.expires_at < cutoff)
                )
            }
            s.sessions = {
                k: r for k, r in s.sessions.items() if not (r.deleted_at is not None and r.deleted_at < cutoff)
            }
            s.leases = {
                k: r
                for k, r in s.leases.items()
                if not ((r.deleted_at is not None and r.deleted_at < cutoff) or r.lease_expires_at < cutoff)
            }
            s.approvals = {
                k: r for k, r in s.approvals.items() if not (r.resolved_at is not None and r.resolved_at < cutoff)
            }
            after = len(s.resume_tokens) + len(s.session_tokens) + len(s.sessions) + len(s.leases) + len(s.approvals)
            return before - after

    async def get_audit_head(self) -> tuple[int, str] | None:
        """Return the in-memory audit-chain head ``(seq, record_hash)``, or
        ``None`` if none recorded yet.  NON-DURABLE: the memory backend loses the
        head on restart (consistent with this backend's documented volatility),
        so cross-restart anti-rollback only holds for the sqlite backend."""
        async with self._lock:
            return self._state.audit_head

    async def set_audit_head(self, seq: int, record_hash: str) -> None:
        """Persist the audit-chain head MONOTONICALLY.

        A lower-or-equal seq is a NO-OP, so the head never moves backwards
        (anti-rollback guard).  NON-DURABLE — see ``get_audit_head``."""
        async with self._lock:
            if self._state.audit_head is not None and self._state.audit_head[0] >= seq:
                return
            self._state.audit_head = (seq, record_hash)

    def session_store(self, tx: MemoryTransaction) -> MemorySessionStore:
        return MemorySessionStore(tx.state, tx)

    def token_store(self, tx: MemoryTransaction) -> MemoryTokenStore:
        return MemoryTokenStore(tx.state, tx)

    def approval_store(self, tx: MemoryTransaction) -> MemoryApprovalStore:
        return MemoryApprovalStore(tx.state, tx)

    def lease_store(self, tx: MemoryTransaction) -> MemoryLeaseStore:
        return MemoryLeaseStore(tx.state, tx)

    def graphical_target_store(self, tx: Any) -> MemoryGraphicalTargetStore:
        return MemoryGraphicalTargetStore(tx.state, tx)

File: packages/provide-uterm-go/controlplane/memory/engine.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package memory

import (
	"context"
	"sync"

	cp "github.com/provide-io/provide-uterm/packages/provide-uterm-go/controlplane"
)

// Engine is the in-memory control-plane backend with shared mutable state. Port
// of control.plane.memory.engine.MemoryControlPlane.
type Engine struct {
	config cp.Config
	caps   cp.EngineCapabilities
	state  *State
	lock   sync.Mutex
	// auditMu guards audit-head access, mirroring Python's use of the same
	// asyncio lock for the head; kept separate here so head reads/writes never
	// contend with transaction commits.
	auditMu sync.Mutex
}

// New builds an in-memory Engine from config. Defaults are applied so a
// zero-value Config yields the Python defaults.
func New(config cp.Config) *Engine {
	config = config.Normalized()
	return &Engine{
		config: config,
		caps:   config.Capabilities,
		state:  newState(),
	}
}

// Capabilities returns the engine feature flags.
func (e *Engine) Capabilities() cp.EngineCapabilities { return e.caps }

// Open is a no-op for the memory backend.
func (e *Engine) Open(context.Context) error { return nil }

// Close is a no-op for the memory backend.
func (e *Engine) Close(context.Context) error { return nil }

// Migrate is a no-op for the memory backend.
func (e *Engine) Migrate(context.Context) error { return nil }

// Begin starts a new snapshot-isolated transaction.
func (e *Engine) Begin(context.Context) (cp.Tx, error) {
	e.lock.Lock()
	defer e.lock.Unlock()
	return newTransaction(e.state, &e.lock), nil
}

// State exposes the shared root state so tests can seed rows directly (mirroring
// the Python tests that reach into “plane._state“).
func (e *Engine) State() *State { return e.state }

// Reap physically drops rows whose soft-delete/expiry timestamp is older than
// now-retentionS (strict <, IS-NOT-NULL guards). Returns the number removed.
// Port of MemoryControlPlane.reap.
func (e *Engine) Reap(_ context.Context, now float64, retentionS int) (int, error) {
	cutoff := now - float64(retentionS)
	e.lock.Lock()
	defer e.lock.Unlock()
	s := e.state
	before := len(s.ResumeTokens) + len(s.SessionTokens) + len(s.Sessions) + len(s.Leases) + len(s.Approvals)

	for k, r := range s.ResumeTokens {
		if (r.RevokedAt.Valid && r.RevokedAt.Float64 < cutoff) || r.ExpiresAt < cutoff {
			delete(s.ResumeTokens, k)
		}
	}
	for k, r := range s.SessionTokens {
		if (r.RevokedAt.Valid && r.RevokedAt.Float64 < cutoff) ||
			(r.ExpiresAt.Valid && r.ExpiresAt.Float64 < cutoff) {
			delete(s.SessionTokens, k)
		}
	}
	for k, r := range s.Sessions {
		if r.DeletedAt.Valid && r.DeletedAt.Float64 < cutoff {
			delete(s.Sessions, k)
		}
	}
	for k, r := range s.Leases {
		if (r.DeletedAt.Valid && r.DeletedAt.Float64 < cutoff) || r.LeaseExpiresAt < cutoff {
			delete(s.Leases, k)
		}
	}
	for k, r := range s.Approvals {
		if r.ResolvedAt.Valid && r.ResolvedAt.Float64 < cutoff {
			delete(s.Approvals, k)
		}
	}

	after := len(s.ResumeTokens) + len(s.SessionTokens) + len(s.Sessions) + len(s.Leases) + len(s.Approvals)
	return before - after, nil
}

// GetAuditHead returns the in-memory audit-chain head, or nil if unset.
// NON-DURABLE: the head is lost on restart.
func (e *Engine) GetAuditHead(context.Context) (*cp.AuditHead, error) {
	e.auditMu.Lock()
	defer e.auditMu.Unlock()
	if e.state.AuditHead == nil {
		return nil, nil
	}
	head := *e.state.AuditHead
	return &head, nil
}

// SetAuditHead persists the head monotonically: a lower-or-equal seq is a no-op.
func (e *Engine) SetAuditHead(_ context.Context, seq int64, recordHash string) error {
	e.auditMu.Lock()
	defer e.auditMu.Unlock()
	if e.state.AuditHead != nil && e.state.AuditHead.Seq >= seq {
		return nil
	}
	e.state.AuditHead = &cp.AuditHead{Seq: seq, RecordHash: recordHash}
	return nil
}

// txState resolves a Tx to a memory transaction's working state, panicking on a
// foreign Tx (the same failure mode as passing the wrong engine's tx in Python).
func txState(tx cp.Tx) *State {
	mt, ok := tx.(*Transaction)
	if !ok {
		panic("memory: transaction is not a *memory.Transaction")
	}
	return mt.state
}

// SessionStore returns a session store bound to tx.
func (e *Engine) SessionStore(tx cp.Tx) cp.SessionStore { return &sessionStore{state: txState(tx)} }

// TokenStore returns a token store bound to tx.
func (e *Engine) TokenStore(tx cp.Tx) cp.TokenStore { return &tokenStore{state: txState(tx)} }

// ApprovalStore returns an approval store bound to tx.
func (e *Engine) ApprovalStore(tx cp.Tx) cp.ApprovalStore { return &approvalStore{state: txState(tx)} }

// LeaseStore returns a lease store bound to tx.
func (e *Engine) LeaseStore(tx cp.Tx) cp.LeaseStore { return &leaseStore{state: txState(tx)} }

// GraphicalTargetStore returns a graphical target store bound to tx.
func (e *Engine) GraphicalTargetStore(tx cp.Tx) cp.GraphicalTargetStore {
	return &graphicalTargetStore{state: txState(tx)}
}
